File: src/game/player_manager.py
"""Player management for Skull King game."""

from typing import List


class PlayerManager:
    """Manages player addition and validation."""
    
    MIN_PLAYERS = 2
    MAX_PLAYERS = 8
    
    def __init__(self):
        """Initialize player manager."""
        self._players: List[str] = []
        self._game_started = False
    
    def add_player(self, name: str) -> None:
        """Add a player with the given name.
        
        Args:
            name: The player's name (must be unique and non-empty)
            
        Raises:
            ValueError: If game has started, name is empty, name is not unique,
                       or max player limit is reached
        """
        if self._game_started:
            raise ValueError("Cannot add players after game has started")
        
        if not name or not name.strip():
            raise ValueError("Player name cannot be empty")
        
        name = name.strip()
        
        if name in self._players:
            raise ValueError(f"Player '{name}' already exists")
        
        if len(self._players) >= self.MAX_PLAYERS:
            raise ValueError(
                f"Maximum {self.MAX_PLAYERS} players allowed"
            )
        
        self._players.append(name)
    
    def remove_player(self, name: str) -> None:
        """Remove a player by name.
        
        Args:
            name: The player's name to remove
            
        Raises:
            ValueError: If game has started or player doesn't exist
        """
        if self._game_started:
            raise ValueError("Cannot remove players after game has started")
        
        if name not in self._players:
            raise ValueError(f"Player '{name}' not found")
        
        self._players.remove(name)
    
    def get_players(self) -> List[str]:
        """Get list of players in order added.
        
        Returns:
            List of player names in order they were added
        """
        return self._players.copy()
    
    def get_player_count(self) -> int:
        """Get number of players.
        
        Returns:
            Number of players currently in game
        """
        return len(self._players)
    
    def has_valid_player_count(self) -> bool:
        """Check if current player count is valid to start game.
        
        Returns:
            True if player count is between MIN_PLAYERS and MAX_PLAYERS
        """
        return self.MIN_PLAYERS <= len(self._players) <= self.MAX_PLAYERS
    
    def can_start_game(self) -> bool:
        """Check if game can be started.
        
        Returns:
            True if game has not started and player count is valid
        """
        return not self._game_started and self.has_valid_player_count()
    
    def start_game(self) -> None:
        """Mark game as started.
        
        Raises:
            ValueError: If not enough players or game already started
        """
        if self._game_started:
            raise ValueError("Game has already started")
        
        if not self.has_valid_player_count():
            raise ValueError(
                f"Need between {self.MIN_PLAYERS} and {self.MAX_PLAYERS} "
                f"players to start game (currently {len(self._players)})"
            )
        
        self._game_started = True
    
    def is_game_started(self) -> bool:
        """Check if game has started.
        
        Returns:
            True if game has been started
        """
        return self._game_started
    
    def reset(self) -> None:
        """Reset player manager to initial state."""
        self._players = []
        self._game_started = False

File: js/playerManager.js
/**
 * PlayerManager - Manages player data and operations for Skull King game
 * Handles adding/removing players, validation, and game readiness checks
 */
class PlayerManager {
  constructor() {
    this.players = [];
    this.MIN_PLAYERS = 2;
    this.MAX_PLAYERS = 8;
  }

  /**
   * Generates a unique ID using timestamp and random value
   * @returns {string} Unique player ID
   */
  generateId() {
    return `player_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Checks if a name already exists in the player list
   * @param {string} name - Player name to check
   * @returns {boolean} True if name exists (case-insensitive)
   */
  nameExists(name) {
    return this.players.some(player => player.name.toLowerCase() === name.toLowerCase());
  }

  /**
   * Adds a new player to the game
   * @param {string} name - Player name (must be unique, 1+ characters)
   * @returns {object} The created player object or error object
   * @throws {Error} If constraints are violated
   */
  addPlayer(name) {
    // Validate name is provided and is a string
    if (!name || typeof name !== 'string') {
      throw new Error('Player name must be a non-empty string');
    }

    // Trim whitespace
    const trimmedName = name.trim();

    if (!trimmedName) {
      throw new Error('Player name cannot be empty or whitespace only');
    }

    // Check if at maximum capacity
    if (this.players.length >= this.MAX_PLAYERS) {
      throw new Error(`Cannot add more than ${this.MAX_PLAYERS} players`);
    }

    // Check for duplicate name
    if (this.nameExists(trimmedName)) {
      throw new Error(`A player named "${trimmedName}" already exists`);
    }

    // Create new player object
    const newPlayer = {
      id: this.generateId(),
      name: trimmedName,
      score: 0
    };

    // Add player to array
    this.players.push(newPlayer);

    return newPlayer;
  }

  /**
   * Removes a player by ID
   * @param {string} id - Player ID to remove
   * @returns {boolean} True if player was removed, false if not found
   */
  removePlayer(id) {
    if (!id || typeof id !== 'string') {
      throw new Error('Invalid player ID');
    }

    const initialLength = this.players.length;
    this.players = this.players.filter(player => player.id !== id);

    return this.players.length < initialLength;
  }

  /**
   * Gets all current players
   * @returns {array} Array of player objects
   */
  getPlayers() {
    // Return a copy to prevent external modifications
    return JSON.parse(JSON.stringify(this.players));
  }

  /**
   * Checks if game can start (minimum 2 players required)
   * @returns {boolean} True if 2 or more players exist
   */
  canStartGame() {
    return this.players.length >= this.MIN_PLAYERS;
  }

  /**
   * Gets the count of current players
   * @returns {number} Number of players
   */
  getPlayerCount() {
    return this.players.length;
  }

  /**
   * Clears all players (useful for game reset)
   */
  clearPlayers() {
    this.players = [];
  }

  /**
   * Gets a player by ID
   * @param {string} id - Player ID
   * @returns {object|null} Player object or null if not found
   */
  getPlayerById(id) {
    const player = this.players.find(p => p.id === id);
    return player ? JSON.parse(JSON.stringify(player)) : null;
  }

  /**
   * Gets a player by name
   * @param {string} name - Player name
   * @returns {object|null} Player object or null if not found
   */
  getPlayerByName(name) {
    const player = this.players.find(p => p.name.toLowerCase() === name.toLowerCase());
    return player ? JSON.parse(JSON.stringify(player)) : null;
  }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
  module.exports = PlayerManager;
}
